src/VisualStudio/Core/Def/Implementation/CodeModel/ICodeModelInstanceFactory.cs

```diff
@@ -0,0 +1,21 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.CodeModel
+{
+    interface ICodeModelInstanceFactory
+    {
+        /// <summary>
+        /// Requests the project system to create a <see cref="EnvDTE.FileCodeModel"/> through the project system.
+        /// </summary>
+        /// <remarks>
+        /// This is sometimes necessary because it's possible to go from one <see cref="EnvDTE.FileCodeModel"/> to another,
+        /// but the language service can't create that instance directly, as it doesn't know what the <see cref="EnvDTE.FileCodeModel.Parent"/>
+        /// member should be. The expectation is the implementer of this will do what is necessary and call back into the code model implementation
+        /// with a parent object.</remarks>
+        EnvDTE.FileCodeModel TryCreateFileCodeModelThroughProjectSystem(string filePath);
```

Comment:
what does this mean for files that are linked and shared across different projects?

Reply:
There is an implicit "project" context that is implied an implementer of this interface has. If you look at the various implementations, you'll see they either go right back to the project system's project-specific type (ICSharpProjectRoot in the csproj.dll case) or they go to a hierarchy for the project we already have, and grab DTE stuff from there.